### PR TITLE
feat: allow passing a command name in autocomplete decorator

### PR DESCRIPTION
**Pull request by V3ntus**
## About

This allows a string to be passed (as per Delta's suggestion) and passed in a cache `get()` call

## Checklist

- [X] I've ran `pre-commit` to format and lint the change(s) made.
- [X] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [X] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent): A comment by [@pwuertz](https://github.com/goverfl0w/discord-interactions/pull/422#issuecomment-1011518065) on #422 
- [X] I've made this pull request for/as: (check all that apply)
  - [X] Documentation
  - [ ] Breaking change
  - [X] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

File: interactions/client.py [hunk 1]
```diff
@@ -553,15 +553,28 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
         _command: Union[Snowflake, int] = (
             command.id if isinstance(command, ApplicationCommand) else command
         )
 
+        if isinstance(command, ApplicationCommand):
+            _command: Union[Snowflake, int] = command.id
+        elif isinstance(command, str):
+            _command_obj = self.http.cache.interactions.get(command)
+            if not _command_obj:
+                self.synchronize()
```
Comment (FayeDel) [inline interactions/client.py]: `self.synchronize()` is a coroutine.

**Review by FayeDel — CHANGES_REQUESTED** [review]

**Comment by V3ntus** [comment]
@DeltaXWizard Oops, missed that

Is this method a good way to ensure callback functionality (calling `synchronize()` when returned None)? If the autocomplete callback is registered before its command counterpart, this would result in a failure. Is that left up to the users to be aware of?

File: interactions/client.py [hunk 2]
```diff
@@ -553,15 +553,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
+            _command_obj = self.http.cache.interactions.get(command)
+            if not _command_obj:
+                _sync_task = ensure_future(self.synchronize(), loop=self.loop)
+                while not _sync_task.done():
+                    pass  # wait for sync to finish
+                _command_obj = self.http.cache.interactions.get(command)
+                if not _command_obj:
+                    raise InteractionException(6, message="The command does not exist")
+            _command: Union[Snowflake, int] = int(_command_obj.id)
+        else:
```
Comment (EdVraz) [inline interactions/client.py:577]: I'd recommend checking if it is actually an integer and if not raising an exception.

File: interactions/client.py [hunk 3]
```diff
@@ -553,15 +553,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
         :type name: str
-        :param command: The command or commnd ID with the option.
-        :type command: Union[ApplicationCommand, int]
+        :param command: The command, command ID, or command name with the option.
+        :type command: Union[ApplicationCommand, int, str]
         :return: A callable response.
         :rtype: Callable[..., Any]
         """
         _command: Union[Snowflake, int] = (
             command.id if isinstance(command, ApplicationCommand) else command
```
Comment (EdVraz) [inline interactions/client.py:562]:
```suggestion
```

File: interactions/client.py [hunk 4]
```diff
@@ -553,15 +553,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
         :param name: The name of the option to autocomplete.
         :type name: str
-        :param command: The command or commnd ID with the option.
-        :type command: Union[ApplicationCommand, int]
+        :param command: The command, command ID, or command name with the option.
+        :type command: Union[ApplicationCommand, int, str]
         :return: A callable response.
         :rtype: Callable[..., Any]
         """
         _command: Union[Snowflake, int] = (
```
Comment (EdVraz) [inline interactions/client.py:561]:
```suggestion
```

File: interactions/client.py [hunk 5]
```diff
@@ -553,15 +553,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
             command.id if isinstance(command, ApplicationCommand) else command
         )
```
Comment (EdVraz) [inline interactions/client.py:563]:
```suggestion
```

Comment (EdVraz) [reply interactions/client.py:563]: remove this, it is iterated twice and unnecessary

File: interactions/client.py (hunk 2 above)
Comment (EdVraz) [inline interactions/client.py:577]:
```suggestion
        elif isinstance(command, int):
        _command: Union[Snowflake, int] = command
```

File: interactions/client.py [hunk 6]
```diff
@@ -553,15 +553,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
+            if not _command_obj:
+                _sync_task = ensure_future(self.synchronize(), loop=self.loop)
+                while not _sync_task.done():
+                    pass  # wait for sync to finish
+                _command_obj = self.http.cache.interactions.get(command)
+                if not _command_obj:
+                    raise InteractionException(6, message="The command does not exist")
+            _command: Union[Snowflake, int] = int(_command_obj.id)
+        else:
+            _command: Union[Snowflake, int] = command
```
Comment (EdVraz) [inline interactions/client.py:578]:
```suggestion

```

File: interactions/client.py [hunk 7]
```diff
@@ -553,15 +553,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
+
```
Comment (EdVraz) [inline interactions/client.py:579]:
```suggestion

else:
       raise ValueError("you provided a wrong data type! Only 'ApplicationCommand' 'str' and 'int' are allowed!")  # TODO: move to custom error formatter
       
```

**Review by sillyElvy — CHANGES_REQUESTED** [review]
this are my first suggestions, sorry if it looks horrible

File: interactions/client.py [hunk 8]
```diff
@@ -555,15 +555,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
```
Comment (sillyElvy) [inline interactions/client.py]:
```suggestion
        elif isinstance(command, int):
            _command: Union[Snowflake, int] = command
        else:
              raise ValueError("You can only insert string, intergers and ApplicationCommands here!") # TODO: move to custom error formatter
```

File: interactions/client.py [hunk 9]
```diff
@@ -555,15 +555,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
-        :param command: The command or commnd ID with the option.
-        :type command: Union[ApplicationCommand, int]
+        :param command: The command, command ID, or command name with the option.
+        :type command: Union[ApplicationCommand, int, str]
         :return: A callable response.
         :rtype: Callable[..., Any]
         """
         _command: Union[Snowflake, int] = (
             command.id if isinstance(command, ApplicationCommand) else command
         )
```
Comment (sillyElvy) [inline interactions/client.py]:
```suggestion
```

**Review by sillyElvy — CHANGES_REQUESTED** [review]
now its better lol

File: interactions/client.py (hunk 8 above)
Comment (sillyElvy) [reply interactions/client.py]: I'd recommend checking if it is actually an integer and if not raising an exception.

File: interactions/client.py (hunk 9 above)
Comment (sillyElvy) [reply interactions/client.py]: remove it, it's a duplicate check

File: interactions/client.py [hunk 10]
```diff
@@ -555,14 +555,30 @@ async def autocomplete_choice_list(ctx, user_input: str = ""):
...
+                    pass  # wait for sync to finish
+                _command_obj = self.http.cache.interactions.get(command)
+                if not _command_obj:
+                    raise InteractionException(6, message="The command does not exist")
+            _command: Union[Snowflake, int] = int(_command_obj.id)
+        elif isinstance(command, int):
+            _command: Union[Snowflake, int] = command
+        else:
+            raise ValueError(
+                "You can only insert string, intergers and ApplicationCommands here!"
```
Comment (sillyElvy) [inline interactions/client.py]:
```suggestion
                "You can only insert strings, integers and ApplicationCommands here!"
```
sry one typo

**Review by sillyElvy — CHANGES_REQUESTED** [review]
Sorry one typo i made myself lol

**Review by sillyElvy — APPROVED** [review]
LGTM :+1:

**Review by FayeDel — APPROVED** [review]
Looked this over, LGTM :)

**Pull request merged**